<a href="https://colab.research.google.com/github/Abhimokashe/ML_Handle_Imbalanced_Credit_Card_Dataset/blob/main/Imbalanced_dataset_Credit_card.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [5]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split
from sklearn.dummy import DummyClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score,f1_score,confusion_matrix,recall_score

In [6]:
df = pd.read_csv('/content/drive/MyDrive/Credit_card_imbalanced.zip')
df

Time         V1         V2        V3        V4        V5  \
0            0.0  -1.359807  -0.072781  2.536347  1.378155 -0.338321   
1            0.0   1.191857   0.266151  0.166480  0.448154  0.060018   
2            1.0  -1.358354  -1.340163  1.773209  0.379780 -0.503198   
3            1.0  -0.966272  -0.185226  1.792993 -0.863291 -0.010309   
4            2.0  -1.158233   0.877737  1.548718  0.403034 -0.407193   
...          ...        ...        ...       ...       ...       ...   
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
0       0.462388  0.239599  0.098698  0.363787  ... -0.018307  0.277838   
1      -0.082361 -0.078803  0.085102 -0.255425  ... -0.225775 -0.638672   
2       1.800499  0.791461  0.247676 -1.514654  ...  0.247998  0.771679   
3       1.247203  0.237609  0.377436 -1.387024  ... -0.108300  0.005274   
4       0.095921  0.592941 -0.270533  0.817739  ... -0.009431  0.798278   
...          ...       ...       ...       ...  ...       ...       ...   
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
0      -0.110474  0.066928  0.128539 -0.189115  0.133558 -0.021053  149.62   
1       0.101288 -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69   
2       0.909412 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752  378.66   
3      -0.190321 -1.175575  0.647376 -0.221929  0.062723  0.061458  123.50   
4      -0.137458  0.141267 -0.206010  0.502292  0.219422  0.215153   69.99   
...          ...       ...       ...       ...       ...       ...     ...   
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
0           0  
1           0  
2           0  
3           0  
4           0  
...       ...  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[284807 rows x 31 columns]

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [8]:
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [9]:
df.describe()

Time            V1            V2            V3            V4  \
count  284807.000000  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean    94813.859575  1.168375e-15  3.416908e-16 -1.379537e-15  2.074095e-15   
std     47488.145955  1.958696e+00  1.651309e+00  1.516255e+00  1.415869e+00   
min         0.000000 -5.640751e+01 -7.271573e+01 -4.832559e+01 -5.683171e+00   
25%     54201.500000 -9.203734e-01 -5.985499e-01 -8.903648e-01 -8.486401e-01   
50%     84692.000000  1.810880e-02  6.548556e-02  1.798463e-01 -1.984653e-02   
75%    139320.500000  1.315642e+00  8.037239e-01  1.027196e+00  7.433413e-01   
max    172792.000000  2.454930e+00  2.205773e+01  9.382558e+00  1.687534e+01   

                 V5            V6            V7            V8            V9  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   9.604066e-16  1.487313e-15 -5.556467e-16  1.213481e-16 -2.406331e-15   
std    1.380247e+00  1.332271e+00  1.237094e+00  1.194353e+00  1.098632e+00   
min   -1.137433e+02 -2.616051e+01 -4.355724e+01 -7.321672e+01 -1.343407e+01   
25%   -6.915971e-01 -7.682956e-01 -5.540759e-01 -2.086297e-01 -6.430976e-01   
50%   -5.433583e-02 -2.741871e-01  4.010308e-02  2.235804e-02 -5.142873e-02   
75%    6.119264e-01  3.985649e-01  5.704361e-01  3.273459e-01  5.971390e-01   
max    3.480167e+01  7.330163e+01  1.205895e+02  2.000721e+01  1.559499e+01   

       ...           V21           V22           V23           V24  \
count  ...  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   ...  1.654067e-16 -3.568593e-16  2.578648e-16  4.473266e-15   
std    ...  7.345240e-01  7.257016e-01  6.244603e-01  6.056471e-01   
min    ... -3.483038e+01 -1.093314e+01 -4.480774e+01 -2.836627e+00   
25%    ... -2.283949e-01 -5.423504e-01 -1.618463e-01 -3.545861e-01   
50%    ... -2.945017e-02  6.781943e-03 -1.119293e-02  4.097606e-02   
75%    ...  1.863772e-01  5.285536e-01  1.476421e-01  4.395266e-01   
max    ...  2.720284e+01  1.050309e+01  2.252841e+01  4.584549e+00   

                V25           V26           V27           V28         Amount  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  284807.000000   
mean   5.340915e-16  1.683437e-15 -3.660091e-16 -1.227390e-16      88.349619   
std    5.212781e-01  4.822270e-01  4.036325e-01  3.300833e-01     250.120109   
min   -1.029540e+01 -2.604551e+00 -2.256568e+01 -1.543008e+01       0.000000   
25%   -3.171451e-01 -3.269839e-01 -7.083953e-02 -5.295979e-02       5.600000   
50%    1.659350e-02 -5.213911e-02  1.342146e-03  1.124383e-02      22.000000   
75%    3.507156e-01  2.409522e-01  9.104512e-02  7.827995e-02      77.165000   
max    7.519589e+00  3.517346e+00  3.161220e+01  3.384781e+01   25691.160000   

               Class  
count  284807.000000  
mean        0.001727  
std         0.041527  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         0.000000  
max         1.000000  

[8 rows x 31 columns]

In [10]:
df.shape

(284807, 31)

In [11]:
# Class Imbalance
df.Class.value_counts()

0    284315
1       492
Name: Class, dtype: int64

In [12]:
# Print percentage of questions where target == 1
(len(df.loc[df.Class==1])) / (len(df.loc[df.Class==0]))*100

0.17304750013189596

# **Create Train and Test Sets**

The training set is used to build and validate model,while test set is reserved for testing the model on unseen data.

In [13]:
# Prepare data for modelling
# Separate input features and target

y = df.Class
X = df.drop('Class',axis=1)

# Setting up training and testing sets
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.25,random_state=27)

Baseline Models

In [14]:
# Dummy classifier to predict only target 0
dummy = DummyClassifier(strategy='most_frequent').fit(X_train,y_train)
dummy_pred = dummy.predict(X_test)

# Checking unique labels
print('Unique predicted labels: ',(np.unique(dummy_pred)))

# Checking accuracy
print('Accuracy Score: ',accuracy_score(y_test,dummy_pred))

Unique predicted labels:  [0]
Accuracy Score:  0.9981461194910255


As predicted our accuracy for classifying all transcations as not fraud is 99.8%!

As the Dummy Classifier predicts only class 0, it is clearly not a good option for our objective of correctly classifying fraudelent transactions.

Let's see how logistic regression performs on this dataset.

In [15]:
# Modeling the data as is
# Train model
lr = LogisticRegression(solver='liblinear').fit(X_train,y_train)

# Predict on training dataset
lr_pred = lr.predict(X_test)

# Checking accuracy score
accuracy_score(y_test,lr_pred)

# Checking unique values
predictions = pd.DataFrame(lr_pred)
predictions[0].value_counts()

0    71108
1       94
Name: 0, dtype: int64

Logistic regression outperformed the Dummy Classifier! We can see that it predicted 94 instances of class 1, so this is definitely improvement. But can we do better?

Let's see if we can apply some techniques for dealing with class imbalance to improve these results.

1. Change the performance metric

Accuracy is not the best metric to use when evaluating imbalanced datasets as it can be misleading. Metrics that can be provide better insight include:

*   Confusion Metrics : A table showing correct predictions and types of incorrect predictions.
*   Precision : The number of true positives divided by all positive predictions. Precision is also called positive Predictive Value. It is measure of a classifier's exactness. Low precision indicates a high number of false positives.
*   Recall : The number of true positives divided by the number of positives values in the test data. Recall is also called Sensitivity or the True Positive Rate. It is a measure of a classifiers completeness. Low recall indicates a high number of false negatives.
*   F1 Score : The weighted average of precision and recall.

Since our main objective with the dataset is to prioritize accurately classifying fraud cases the recall score can be considered our main metric to use for evaluating outcomes. 





In [16]:
# F1 Score
f1_score(y_test,lr_pred)

0.7522123893805309

In [17]:
# Confusion matrix
pd.DataFrame(confusion_matrix(y_test,lr_pred))

0   1
0  71061   9
1     47  85

We have a very high accuracy score of 0.999 but a F1 score of only 0.752. And from the confusion matrix, we can see we are misclassifying several observations leading to a recall score of only 0.64.

# **2. Change of Algorithm**

While in every machine learning problem, its a good rule of thumb to try a variety of algorithms, it can be especially beneficial with imbalanced datasets. Decision trees frequently perform well on imbalanced data. They work by learning a hierachy of if/else questions. This can force both classes to be addressed.

In [18]:
from sklearn.ensemble import RandomForestClassifier

# train model
rfc = RandomForestClassifier(n_estimators = 10).fit(X_train,y_train)

# predict on test set
rfc_pred = rfc.predict(X_test)

accuracy_score(y_test,rfc_pred)

0.9995786635206876

In [19]:
# f1 score
f1_score(y_test,rfc_pred)

0.8739495798319327

In [20]:
# Confusion matrix
pd.DataFrame(confusion_matrix(y_test,rfc_pred))

0    1
0  71068    2
1     28  104

In [21]:
# Recall score
recall_score(y_test,rfc_pred)

0.7878787878787878

# **3. Oversampling Minority Class**

Oversampling can be defined as adding more copies minority class. Oversampling can be good choice when you don't have a ton data to work with. A con to consider when undersampling is that it can cause overfitting and poor generalization to your test set.

We will use the resampling module from Scikit-Learn to randomly replicate samples from the minority class.


**Important Note**

Always split into test and train sets BEFORE trying any resampling techniques! Oversampling before splitting the data can allow the exact same observations to be present in both the test and train sets! This can allow our model to simply memorize specific data points and cause overfitting.

Note

The simplest implementation of over-sampling is to duplicate random records from the minority class, which can cause overfishing.

In under-sampling, the simplest technique involves removing random records from the majority class, which can cause loss of information.




In [22]:
from sklearn.metrics import classification_report
from sklearn.datasets import make_classification
from imblearn.over_sampling import RandomOverSampler
from imblearn.under_sampling import RandomUnderSampler
from collections import Counter

In [41]:
 # defining the dataset
 X1, y1 = make_classification(n_samples= 284807, weights=[.50])

 # class distribution
 print(Counter(y1))

Counter({1: 142446, 0: 142361})


In [50]:
y.value_counts()

0    284315
1       492
Name: Class, dtype: int64

In [49]:
y_train_df.value_counts()

Class
0        213236
1           369
dtype: int64

In [24]:
# Prepare data for modelling
# Separate input features and target

y = df.Class
X = df.drop('Class',axis=1)

# Setting up training and testing sets
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.25,stratify=y,random_state=27)

In [25]:
# instantiating the random over sampler 
ros = RandomOverSampler()

# resampling X, y
X_ros, y_ros = ros.fit_resample(X_train, y_train)

# new class distribution 
print(Counter(y_ros))

Counter({0: 213236, 1: 213236})


In [26]:

# Training with imbalance data
classifier = LogisticRegression(class_weight='balanced',solver='newton-cg')
model = classifier.fit(X_ros,y_ros)

print(classification_report(y_test, model.predict(X_test)))

              precision    recall  f1-score   support

           0       1.00      0.98      0.99     71079
           1       0.07      0.93      0.12       123

    accuracy                           0.98     71202
   macro avg       0.53      0.96      0.55     71202
weighted avg       1.00      0.98      0.99     71202



/usr/local/lib/python3.9/dist-packages/sklearn/utils/optimize.py:210: ConvergenceWarning: newton-cg failed to converge. Increase the number of iterations.
  warnings.warn(


In [51]:
y_pred_train = classifier.predict(X_ros)

In [53]:
accuracy_score(y_pred_train,y_ros)

0.9486461010335966

In [64]:
accuracy_score(model.predict(X_test),y_test)

0.9995505744220669

In [54]:
recall_score(y_pred_train,y_ros)

0.9739886540986449

In [30]:
# instantiating the random over sampler 
rus = RandomUnderSampler()

# resampling X, y
X_rus, y_rus = rus.fit_resample(X_train, y_train)

# new class distribution 
print(Counter(y_rus))

Counter({0: 369, 1: 369})


In [31]:
print(classification_report(y_test,model.predict(X_test)))

              precision    recall  f1-score   support

           0       1.00      0.98      0.99     71079
           1       0.07      0.93      0.12       123

    accuracy                           0.98     71202
   macro avg       0.53      0.96      0.55     71202
weighted avg       1.00      0.98      0.99     71202



In [55]:
accuracy_score(model.predict(X_rus),y_rus)

0.9512195121951219

In [61]:
accuracy_score(model.predict(X_test),y_test)

0.9995505744220669

In [32]:
print(classification_report(y_rus,model.predict(X_rus)))

              precision    recall  f1-score   support

           0       0.93      0.98      0.95       369
           1       0.98      0.92      0.95       369

    accuracy                           0.95       738
   macro avg       0.95      0.95      0.95       738
weighted avg       0.95      0.95      0.95       738



In [56]:
from imblearn.over_sampling import SMOTE

# Separate input features and target
y = df.Class
X = df.drop('Class', axis=1)

# setting up testing and training sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=27)

sm = SMOTE()
X_train1, y_train1 = sm.fit_resample(X_train, y_train)

In [57]:
model = rfc.fit(X_train1,y_train1)

In [58]:
y_pred1 = model.predict(X_train1)

In [59]:
accuracy_score(y_pred1,y_train1)

0.999997655279139

In [60]:
y_pred_test = model.predict(X_test)
accuracy_score(y_pred_test,y_test)

0.9995505744220669